gitpod-io/gitpod
# Add e2e tests with Cypress

## Description

🌲 This adds Cypress and some first e2e tests (let's talk about how many we want to add).

🍊🍊🍊🍊🍊🍊🍊🍊🍊🍊🍊🍊🍊
**IMPORTANT! See TODO below!** 
🍊🍊🍊🍊🍊🍊🍊🍊🍊🍊🍊🍊🍊

### 🤔 Reasoning

Apart from the obvious "e2e tests - especially with Cypress - are good for any web application" thing, I wanted to add some tests before tackling the warnings & C. [@gtsiolis and I discussed about on Discord](https://discord.com/channels/816244985187008514/914836798830690344/930546351698247690).

🍩  I'd feel way more comfortable in changing some of the things that need to be changed if I had some tests helping me find any obvious problems deriving from my fixes.

### 📣 Current status

So far I managed to install Cypress, have it run both headless (for CI) and headed (for debugging and implementation of test cases), and I started writing the first test cases.

I also activated TypeScript support and installed Cypress Testing Library (CTL).

You may also find out I updated a `div` to `nav` because it was the semantic thing to do, and because it helped me write a better test case with CTL.

And I updated the service mock with the simplest implementation of two missing methods.

### Updated TODO

- [x] Rebase from main, update test spec (See https://github.com/gitpod-io/gitpod/pull/7611#issuecomment-1014470441)
- [x] Fix conflicts: I know how to deal with `App.tsx`, I'll need some advice @geropl for the ones specifying the docker image (See https://github.com/gitpod-io/gitpod/pull/7611#discussion_r788202086)
- [x] Add quick and easy instructions about tests to README (See again https://github.com/gitpod-io/gitpod/pull/7611#discussion_r788202086 for an interesting gotcha)
- [x] Restore `craco.config.js`
- [x] Reorganize commits (interactive rebase to properly group commits together)

### ⚠️ Original TODO

- Decide where to stop with this first step (MVC)
- Write more tests
- Ask for guidance about properly updating the docker image and the structural changes at large (I basically hacked together things by @mikenikles in a way that's temporary to say the least)
- Find a way to test the unauthenticated scenario (log in page etc)
- Find a way to handle `craco.config.js` so that when running in CI you don't have to customize it (Nx has environments and environment bound configurations, but at this time I don't yet know about craco / cra)
- Update dashboard README with some info about how to write, run, and maintain e2e tests
- Polish

## How to test


- go inside the dashboard directory: `cd components/dashboard`
- in one terminal, launch the dashboard app: `yarn start`
- in another terminal, launch the tests: `yarn cypress install && yarn test:e2e` (re-installation of cypress is something I have to get rid of)
- open port `6080` in the browser to see Cypress via VNC

| Cypress via VNC |
|---|
|![image](https://user-images.githubusercontent.com/462705/149536308-9f50e624-57d0-4a3b-ad4a-5f8b88b7eb22.png)|

- find and click on `app.spec.ts` (the only spec file **at the time of writing**)
- tests should run and pass


## Release Notes

```release-note
NONE
```
